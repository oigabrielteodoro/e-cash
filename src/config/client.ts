export const baseURL = import.meta.env.VITE_API_URL as string
