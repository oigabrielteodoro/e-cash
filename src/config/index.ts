export * from './theme'
export * from './client'
