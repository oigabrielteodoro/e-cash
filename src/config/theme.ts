export const theme = {
  grid: {
    sideBar: {
      open: '25rem',
      closed: '10rem',
    },
  },
  breakpoints: {
    xsm: '320px',
    sm: '425px',
    md: '768px',
    lg: '1440px',
    xlg: '1920px',
    greaterLg: '1441px',
  },
  radius: {
    card: '0.5rem',
    popover: '1rem',
    modal: '1rem',
  },
  font: {
    family: 'Inter',
    sizes: {
      disclaimer: '1.2rem',
      small: '1.4rem',
      paragraph: '1.6rem',
      subtitle: '2.2rem',
      title: '3.2rem',
    },
  },
  layers: {
    base: 1,
    alwaysOnTop: 2,
  },
  colors: {
    white: '#ffffff',
    green: {
      100: '#ecf7ee',
      200: '#c8e2cd',
      500: '#5fb263',
    },
    blue: {
      100: '#e7eff9',
      200: '#b9cfe6',
      300: '#a4b7eA',
      500: '#3a67e1',
      900: '#060c30',
    },
    red: {
      100: '#faede9',
      200: '#eccbc0',
      400: '#fc6471',
      500: '#db5a3a',
    },
    yellow: {
      100: '#fef8eb',
      200: '#f4e1c0',
      500: '#e29936',
    },
    neutral: {
      100: '#eef1f5',
      200: '#f5f5f5',
      300: '#cfd7e3',
      400: '#afbdd0',
      500: '#a1a7ae',
      600: '#4d4f51',
      700: '#393e48',
      900: '#18171d',
    },
  },
  shadow: {
    blue: {
      300: '#a4b7ea30',
    },
    red: {
      200: '#eccbc030',
      400: '#fc647120',
    },
    neutral: {
      500: '#a1a7ae20',
    },
  },
  dropShadow: {
    popover: '0 0 10rem 1rem rgba(0, 0, 0, 0.1)',
  },
}
