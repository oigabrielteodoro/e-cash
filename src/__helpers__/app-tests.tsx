import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
