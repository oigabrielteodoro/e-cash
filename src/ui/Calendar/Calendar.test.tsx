import React from 'react'
import { addMonths, addYears, format, subMonths, subYears } from 'date-fns'

import { render, screen, userEvent, waitFor } from '__helpers__/app-tests'

import { theme } from 'config'

import { Calendar } from '.'

describe('Calendar', () => {
  it('should be able render correctly', async () => {
    const date = new Date()
    const dateFormatted = format(date, 'dd MMM, yyyy')

    render(<Calendar />)

    expect(
      screen.getByRole('button', {
        name: dateFormatted,
      }),
    ).toBeInTheDocument()
  })

  it('should be able open/close calendar', async () => {
    const date = new Date()
    const monthFormatted = format(date, 'MMMM yyyy')
    const dateFormatted = format(date, 'dd MMM, yyyy')

    render(<Calendar />)

    const button = screen.getByRole('button', {
      name: dateFormatted,
    })

    expect(button).toBeInTheDocument()

    // Open calendar
    userEvent.click(button)

    expect(await screen.findByText(monthFormatted)).toBeInTheDocument()

    // Close calendar
    userEvent.click(button)

    await waitFor(() =>
      expect(screen.queryByText(monthFormatted)).not.toBeInTheDocument(),
    )
  })

  it('should be able select single dates', async () => {
    const date = new Date()
    const monthFormatted = format(date, 'MMMM yyyy')
    const dateFormatted = format(date, 'dd MMM, yyyy')

    render(<Calendar />)

    const button = screen.getByRole('button', {
      name: dateFormatted,
    })

    expect(button).toBeInTheDocument()

    // Open calendar
    userEvent.click(button)

    expect(await screen.findByText(monthFormatted)).toBeInTheDocument()

    // Remove default selected date
    userEvent.click(
      await screen.findByRole('button', {
        name: date.getDate().toString(),
      }),
    )

    const singleDate = screen.getByRole('button', {
      name: '1',
    })

    // Select single date
    userEvent.click(singleDate)

    expect(
      await screen.findByRole('button', {
        name: '1',
      }),
    ).toHaveStyle(`background: ${theme.colors.blue[500]}`)
  })

  it('should be able select between dates', async () => {
    const date = new Date()
    const monthFormatted = format(date, 'MMMM yyyy')
    const dateFormatted = format(date, 'dd MMM, yyyy')

    render(<Calendar />)

    const button = screen.getByRole('button', {
      name: dateFormatted,
    })

    expect(button).toBeInTheDocument()

    // Open calendar
    userEvent.click(button)

    expect(await screen.findByText(monthFormatted)).toBeInTheDocument()

    const initialDate = screen.getByRole('button', {
      name: date.getDate().toString(),
    })

    const firstDate = screen.getByRole('button', {
      name: '1',
    })

    const endDate = screen.getByRole('button', {
      name: '4',
    })

    // Unselect initial date
    userEvent.click(initialDate)

    expect(
      await screen.findByRole('button', {
        name: date.getDate().toString(),
      }),
    ).toHaveStyle(`background: ${theme.colors.neutral[100]}`)

    // Select two dates
    userEvent.click(firstDate)
    userEvent.click(endDate)

    expect(
      await screen.findByRole('button', {
        name: '1',
      }),
    ).toHaveStyle(`background: ${theme.colors.blue[500]}`)
    expect(
      await screen.findByRole('button', {
        name: '2',
      }),
    ).toHaveStyle(`background: ${theme.colors.blue[100]}`)
    expect(
      await screen.findByRole('button', {
        name: '3',
      }),
    ).toHaveStyle(`background: ${theme.colors.blue[100]}`)
    expect(
      await screen.findByRole('button', {
        name: '4',
      }),
    ).toHaveStyle(`background: ${theme.colors.blue[500]}`)
  })

  it('should be able reset dates when user selects more than allowed', async () => {
    const date = new Date()
    const monthFormatted = format(date, 'MMMM yyyy')
    const dateFormatted = format(date, 'dd MMM, yyyy')

    render(<Calendar />)

    const button = screen.getByRole('button', {
      name: dateFormatted,
    })

    expect(button).toBeInTheDocument()

    // Open calendar
    userEvent.click(button)

    expect(await screen.findByText(monthFormatted)).toBeInTheDocument()

    const initialDate = screen.getByRole('button', {
      name: date.getDate().toString(),
    })

    const firstDate = screen.getByRole('button', {
      name: '1',
    })

    const endDate = screen.getByRole('button', {
      name: '4',
    })

    // Unselect initial date
    userEvent.click(initialDate)

    expect(
      await screen.findByRole('button', {
        name: date.getDate().toString(),
      }),
    ).toHaveStyle(`background: ${theme.colors.neutral[100]}`)

    // Select two dates
    userEvent.click(firstDate)
    userEvent.click(endDate)

    expect(
      await screen.findByRole('button', {
        name: '1',
      }),
    ).toHaveStyle(`background: ${theme.colors.blue[500]}`)
    expect(
      await screen.findByRole('button', {
        name: '2',
      }),
    ).toHaveStyle(`background: ${theme.colors.blue[100]}`)
    expect(
      await screen.findByRole('button', {
        name: '3',
      }),
    ).toHaveStyle(`background: ${theme.colors.blue[100]}`)
    expect(
      await screen.findByRole('button', {
        name: '4',
      }),
    ).toHaveStyle(`background: ${theme.colors.blue[500]}`)

    const newDateToReset = screen.getByRole('button', {
      name: '6',
    })

    // Select new date to reset previous ones
    userEvent.click(newDateToReset)

    expect(
      await screen.findByRole('button', {
        name: '1',
      }),
    ).not.toHaveStyle(`background: ${theme.colors.blue[500]}`)
    expect(
      await screen.findByRole('button', {
        name: '2',
      }),
    ).not.toHaveStyle(`background: ${theme.colors.blue[100]}`)
    expect(
      await screen.findByRole('button', {
        name: '3',
      }),
    ).not.toHaveStyle(`background: ${theme.colors.blue[100]}`)
    expect(
      await screen.findByRole('button', {
        name: '4',
      }),
    ).not.toHaveStyle(`background: ${theme.colors.blue[500]}`)

    expect(
      await screen.findByRole('button', {
        name: '6',
      }),
    ).toHaveStyle(`background: ${theme.colors.blue[500]}`)
  })

  it('should be able to show next month', async () => {
    const date = new Date()
    const nextDate = addMonths(date, 1)

    const monthFormatted = format(date, 'MMMM yyyy')
    const dateFormatted = format(date, 'dd MMM, yyyy')
    const nextDateFormatted = format(nextDate, 'MMMM yyyy')

    render(<Calendar />)

    const button = screen.getByRole('button', {
      name: dateFormatted,
    })

    expect(button).toBeInTheDocument()

    // Open calendar
    userEvent.click(button)

    expect(await screen.findByText(monthFormatted)).toBeInTheDocument()

    const nextButton = screen.getByRole('button', {
      name: 'Next month',
    })

    // Click in next month button
    userEvent.click(nextButton)

    expect(await screen.findByText(nextDateFormatted)).toBeInTheDocument()
  })

  it('should be able to show next year', async () => {
    const date = new Date()
    const nextDate = addYears(date, 1)

    const monthFormatted = format(date, 'MMMM yyyy')
    const dateFormatted = format(date, 'dd MMM, yyyy')
    const nextDateFormatted = format(nextDate, 'MMMM yyyy')

    render(<Calendar />)

    const button = screen.getByRole('button', {
      name: dateFormatted,
    })

    expect(button).toBeInTheDocument()

    // Open calendar
    userEvent.click(button)

    expect(await screen.findByText(monthFormatted)).toBeInTheDocument()

    const nextButton = screen.getByRole('button', {
      name: 'Next year',
    })

    // Click in next year button
    userEvent.click(nextButton)

    expect(await screen.findByText(nextDateFormatted)).toBeInTheDocument()
  })

  it('should be able to show previous month', async () => {
    const date = new Date()
    const previousDate = subMonths(date, 1)

    const monthFormatted = format(date, 'MMMM yyyy')
    const dateFormatted = format(date, 'dd MMM, yyyy')
    const previousDateFormatted = format(previousDate, 'MMMM yyyy')

    render(<Calendar />)

    const button = screen.getByRole('button', {
      name: dateFormatted,
    })

    expect(button).toBeInTheDocument()

    // Open calendar
    userEvent.click(button)

    expect(await screen.findByText(monthFormatted)).toBeInTheDocument()

    const previousButton = screen.getByRole('button', {
      name: 'Previous month',
    })

    // Click in previous month button
    userEvent.click(previousButton)

    expect(await screen.findByText(previousDateFormatted)).toBeInTheDocument()
  })

  it('should be able to show previous year', async () => {
    const date = new Date()
    const previousDate = subYears(date, 1)

    const monthFormatted = format(date, 'MMMM yyyy')
    const dateFormatted = format(date, 'dd MMM, yyyy')
    const previousDateFormatted = format(previousDate, 'MMMM yyyy')

    render(<Calendar />)

    const button = screen.getByRole('button', {
      name: dateFormatted,
    })

    expect(button).toBeInTheDocument()

    // Open calendar
    userEvent.click(button)

    expect(await screen.findByText(monthFormatted)).toBeInTheDocument()

    const previousButton = screen.getByRole('button', {
      name: 'Previous year',
    })

    // Click in previous year button
    userEvent.click(previousButton)

    expect(await screen.findByText(previousDateFormatted)).toBeInTheDocument()
  })
})
