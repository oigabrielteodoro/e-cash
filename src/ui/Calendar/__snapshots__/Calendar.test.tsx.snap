// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar should be able match snapshot 1`] = `
.c0 {
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: 0;
  color: #a1a7ae;
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c1 svg:last-child {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c1:hover {
  color: #393e48;
}

.c1 > span {
  margin: 0 1rem;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
        />
      </svg>
      <span>
        28 Feb, 2022
      </span>
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="6 9 12 15 18 9"
        />
      </svg>
    </button>
  </div>
  <div
    class="Toastify"
  />
</div>
`;
